### PR TITLE
Modify do_action so default ctrl-click opens tab

### PR DESCRIPTION
**Pull request by azul**
Rebase and whitespace fix of #16530 

New pull request to see if build succeeds.

### CODE REVIEW

**Comment by azul** [comment]
tests are failing as they were before.
This error seems to be related to the changes:
```
undefined is not an object (evaluating 'window.FileActions.currentFile = item.find('td')') in apps/files/js/filelist.js (line 933)
```

File: apps/files/js/filelist.js [hunk 1]
```diff
@@ -933,6 +926,28 @@
 			}
 		},
 
+		openAction: function(id) {
+			var item = this.$fileList.children().filterAttr('data-id', '' + id);
+			var filename = item.attr('data-file');
+			// also set on global object for legacy apps
+			window.FileActions.currentFile = item.find('td');
```
Comment (azul) [inline apps/files/js/filelist.js]: This is the line triggering the `undefined is not an Object` error.
I suspect that item is not set for some reason.

Comment (azul) [reply apps/files/js/filelist.js]: looks more like `window.FileActions ` does not exist anymore at all.
Dropping this.

**Comment by azul** [comment]
Pushed a change that will hopefully help with the tests. I noticed some other strange behavior though...
shift-clicking any file will select all files up to that one plus the last one in the list.
shift-clicking any file in my file browser will select only that file and then start selecting from there with the next shift-click.

I also commented on the initial issue since i don't quite agree with the goal here. But whatever... it's an exercise for now.

**Review by jancborchardt — APPROVED** [review]
@azul I commented in the issue to clarify: Yes, ctrl-click should be the "Open in new tab" action. :)

File: apps/files_sharing/js/public.js [hunk 2]
```diff
@@ -204,10 +204,6 @@ OCA.Sharing.PublicApp = {
 				var $tr = OCA.Files.FileList.prototype._createRow.apply(this, arguments);
 				if (hideDownload === 'true') {
 					this.fileActions.currentFile = $tr.find('td');
-					var mime = this.fileActions.getCurrentMimeType();
-					var type = this.fileActions.getCurrentType();
-					var permissions = this.fileActions.getCurrentPermissions();
-					var action = this.fileActions.getDefault(mime, type, permissions);
```
Comment (azul) [inline apps/files_sharing/js/public.js:210]: none of these were used anywhere.

File: apps/files_trashbin/tests/js/filelistSpec.js [hunk 3]
```diff
@@ -180,7 +180,8 @@ describe('OCA.Trashbin.FileList tests', function () {
 			expect($tr.attr('data-permissions')).toEqual('9'); // read and delete
 			expect($tr.attr('data-mime')).toEqual('text/plain');
 			expect($tr.attr('data-mtime')).toEqual('11111000');
-			expect($tr.find('a.name').attr('href')).toEqual('#');
+			expect($tr.find('a.name').attr('href'))
+				.toEqual(OC.getRootPath() + '/index.php/apps/files?view=trashbin&dir=/&editfile=1');
```
Comment (azul) [inline apps/files_trashbin/tests/js/filelistSpec.js:184]: I'm not sure if this is the desired behavior.
This is in the trashcan. The href used to be `#`.
It looks like now one could open files in the trashcan in another tab. However this does not happen for other reasons.

- [ ] make the trashcan url a `#` again

**Comment by azul** [comment]
Somehow this seems to break the right click menu for me.

**Comment by azul** [comment]
Failing tests seem to be the same as in master.
I am also not seeing the right click menu anymore in master. So i also think that is unrelated. Hope that's an actual change... never liked it anyway :wink:
@juliushaertl @gary-kim I think this one is ready to be merged.

**Comment by gary-kim** [comment]
> I am also not seeing the right click menu anymore in master.

The right click menu is added by the [files_rightclick](https://github.com/nextcloud/files_rightclick). I you want to test it on current master with right click, you will need to add the app and enable it.

**Review by gary-kim — COMMENTED** [review]
The action that occurs from clicking on files to download seems to be broken now. Could you take a look at that?

**Comment by azul** [comment]
> The right click menu is added by the [files_rightclick](https://github.com/nextcloud/files_rightclick) app. If you want to test it on current master with right click, you will need to add the app and enable it.

Ahh.. got it. thanks! :smile:

**Comment by azul** [comment]
> The action that occurs from clicking on files to download seems to be broken now. Could you take a look at that?

Should be fixed.

This is getting close. I would like to add a test for the new urls to make sure they actually trigger the action. Will probably look into that tomorrow.

File: apps/files_sharing/js/public.js [hunk 4]
```diff
@@ -113,8 +113,8 @@ OCA.Sharing.PublicApp = {
 			// Show file preview if previewer is available, images are already handled by the template
 			if (mimetype.substr(0, mimetype.indexOf('/')) !== 'image' && $('.publicpreview').length === 0) {
 				// Trigger default action if not download TODO
-				var action = FileActions.getDefault(mimetype, 'file', OC.PERMISSION_READ);
-				if (typeof action !== 'undefined') {
+				var action = FileActions.getDefaultFileAction(mimetype, 'file', OC.PERMISSION_READ);
+				if (action && action.action) {
 					action($('#filename').val());
```
Comment (gary-kim) [inline apps/files_sharing/js/public.js:118]: Is this supposed to be `action.action($('#filename').val());`?

**Review by gary-kim — COMMENTED** [review]
Noticed a couple more things.
* `Ctrl+click` on a directory in a public share doesn't go into that directory and opens a new tab with the same page.
* `Ctrl+click` on a file in a public share to be opened by the `text` app doesn't work as expected. This is more of an edge case though so I'd say it'd be okay to fix this in a follow up.

File: apps/files_sharing/js/public.js (hunk 4 above)
Comment (azul) [reply apps/files_sharing/js/public.js]: yes... thanks! fixed and renamed the actionSpec from `action` to `spec` so it becomes less confusing.

**Comment by azul** [comment]
Found another issue...
If the file that the link points to has been moved from the folder it will trigger a download of the zipped folder content instead.

@jancborchardt what would the desired behavior be when:
* I create `MyFile.md`.
* I copy the link to `MyFile.md` in the folder (that is used to open it in another tab)
* I move `MyFile.md` to another place
* I click on the link.

I see a number of options:
1. open the same file in the new folder (we can tell where it is by the file id)
2. display a message that the file has been moved (maybe including the new place)
3. only display the folder

I guess 2. would be the right thing to do. But maybe i am missing something.

**Comment by azul** [comment]
> Noticed a couple more things.
> 
>     * `Ctrl+click` on a directory in a public share doesn't go into that directory and opens a new tab with the same page.
> 
>     * `Ctrl+click` on a file in a public share to be opened by the `text` app doesn't work as expected. This is more of an edge case though so I'd say it'd be okay to fix this in a follow up.

I was able to confirm this and track it down a bit.
The call to `linkTo` is called from https://github.com/nextcloud/server/pull/20185/files#diff-dafefd9519924bec4efa9f5af4af8e68L1532 with the dir param as i would expect it.

However on public shares this is handled by 
```
# /apps/files_sharing/js/public.js
this.fileList.linkTo = function (dir) {
  return OC.generateUrl('/s/' + token + '', {dir: dir});
}
```

Turns out this completely ignores the dir parameter. I have no idea why.

@gary-kim do you have a clue?

**Comment by azul** [comment]
58a87d0babcb91aab75b45e630d2fc2fee15691e was the last commit that touched the `linkTo` function in public.js.

**Comment by azul** [comment]
> Noticed a couple more things.
> 
>     * `Ctrl+click` on a directory in a public share doesn't go into that directory and opens a new tab with the same page.
> 
>     * `Ctrl+click` on a file in a public share to be opened by the `text` app doesn't work as expected. This is more of an edge case though so I'd say it'd be okay to fix this in a follow up.

Okay... i figured it out and updated the pull request accordingly. The `linkTo` function in `public.js` had to be [fixed](https://github.com/nextcloud/server/pull/20185/files#diff-f2e13306e758275374ea68815da8af04L254).

**Comment by azul** [comment]
I renamed the `editfile` parameter to `openfile` because depending on the permissions the file will also be shown read only if needed.

@gary-kim I would say this is ready to be merged. I'd be okay with adding more tests but i'd prefer to just see this included :smile: .

In terms of tests i think it would be nice to test that the generated urls actually do what they are supposed to. I really like that the generated url with `openfile` and `path` params on public shares work.

However i have no idea about how to write such a test. If you can point me to an example i'd also be happy to take it from there.

**Review by gary-kim — APPROVED** [review]
LGTM at this point

**Pull request merged**